File: android/src/main/java/com/coeps/capacitor/plugin/ioscookiemover/CookieMoverPlugin.java
package com.coeps.capacitor.plugin.ioscookiemover;

import android.util.Log;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CookieMover")
public class CookieMoverPlugin extends Plugin {

    @PluginMethod
    public void syncNsCookiesToWkCookieStore(PluginCall call) {
        JSObject options = call.getData();

        Log.d(
            "CookieMoverPlugin",
            String.format("syncNsCookiesToWkCookieStore(options: %s) is not implemented in ANDROID plugin!", options)
        );

        JSObject result = new JSObject();
        JSArray preActionCookies = new JSArray();
        JSArray postActionCookies = new JSArray();
        result.put("preActionCookies", preActionCookies);
        result.put("postActionCookies", postActionCookies);
        call.resolve(result);
    }

    @PluginMethod
    public void syncWkCookiesToNsCookieStore(PluginCall call) {
        JSObject options = call.getData();

        Log.d(
            "CookieMoverPlugin",
            String.format("syncWkCookiesToNsCookieStore(options: %s) is not implemented in ANDROID plugin!", options)
        );

        JSObject result = new JSObject();
        JSArray preActionCookies = new JSArray();
        JSArray postActionCookies = new JSArray();
        result.put("preActionCookies", preActionCookies);
        result.put("postActionCookies", postActionCookies);
        call.resolve(result);
    }
}

File: ios/Plugin/CookieMoverPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CookieMoverPlugin)
public class CookieMoverPlugin: CAPPlugin {
    private let implementation = CookieMover()

    @objc func syncNsCookiesToWkCookieStore(_ call: CAPPluginCall) {
        let overwrite = call.getBool("overwrite") ?? false

        print("CookieMoverPlugin - syncNsCookiesToWkCookieStore({ overwrite: \(overwrite) })")

        let callResult = implementation.syncNsCookiesToWkCookieStore(overwrite)

        var result = JSObject()

        var preActionCookies = JSArray()
        var postActionCookies = JSArray()

        for cookie in callResult.preActionCookies {
            preActionCookies.append(JSObject(_immutableCocoaDictionary: cookie))
        }
        for cookie in callResult.postActionCookies {
            postActionCookies.append(JSObject(_immutableCocoaDictionary: cookie))
        }

        result["preActionCookies"] = preActionCookies
        result["postActionCookies"] = postActionCookies

        call.resolve(result)
    }

    @objc func syncWkCookiesToNsCookieStore(_ call: CAPPluginCall) {
        let overwrite = call.getBool("overwrite") ?? false

        print("CookieMoverPlugin - syncWkCookiesToNsCookieStore({ overwrite: \(overwrite) })")

        let callResult = implementation.syncWkCookiesToNsCookieStore(overwrite)

        var result = JSObject()

        var preActionCookies = JSArray()
        var postActionCookies = JSArray()

        for cookie in callResult.preActionCookies {
            preActionCookies.append(JSObject(_immutableCocoaDictionary: cookie))
        }
        for cookie in callResult.postActionCookies {
            postActionCookies.append(JSObject(_immutableCocoaDictionary: cookie))
        }

        result["preActionCookies"] = preActionCookies
        result["postActionCookies"] = postActionCookies

        call.resolve(result)
    }
}
